src/Compilers/CSharp/Test/Syntax/Syntax/SyntaxNodeTests.cs

```diff
@@ -469,20 +469,55 @@ class D
                     """));
             }
 
-            void testContainsHelper2(SyntaxKind[] directiveKinds, CompilationUnitSyntax compilationUnit)
+            static void testContainsHelper2(SyntaxKind[] directiveKinds, CompilationUnitSyntax compilationUnit)
             {
                 Assert.True(compilationUnit.ContainsDirectives);
                 foreach (var directiveKind in directiveKinds)
                     Assert.True(compilationUnit.ContainsDirective(directiveKind));
 
-                for (var kind = SyntaxKind.TildeToken; kind < SyntaxKind.ScopedType; kind++)
+                for (var kind = SyntaxKind.TildeToken; kind < SyntaxKind.XmlElement; kind++)
                 {
                     if (!directiveKinds.Contains(kind))
                         Assert.False(compilationUnit.ContainsDirective(kind));
                 }
             }
         }
 
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/75583")]
+        public void TestContainsDirective_IfIf()
+        {
+            var compilationUnit = SyntaxFactory.ParseCompilationUnit("""
+                if (#if)
+                """);
+            compilationUnit.GetDiagnostics().Verify(
+                // (1,5): error CS1040: Preprocessor directives must appear as the first non-whitespace character on a line
+                // if (#if)
+                TestBase.Diagnostic(ErrorCode.ERR_BadDirectivePlacement, "#").WithLocation(1, 5),
+                // (1,8): error CS1517: Invalid preprocessor expression
+                // if (#if)
+                TestBase.Diagnostic(ErrorCode.ERR_InvalidPreprocExpr, ")").WithLocation(1, 8),
+                // (1,8): error CS1025: Single-line comment or end-of-line expected
+                // if (#if)
+                TestBase.Diagnostic(ErrorCode.ERR_EndOfPPLineExpected, ")").WithLocation(1, 8),
+                // (1,9): error CS1733: Expected expression
+                // if (#if)
+                TestBase.Diagnostic(ErrorCode.ERR_ExpressionExpected, "").WithLocation(1, 9),
+                // (1,9): error CS1026: ) expected
+                // if (#if)
+                TestBase.Diagnostic(ErrorCode.ERR_CloseParenExpected, "").WithLocation(1, 9),
+                // (1,9): error CS1733: Expected expression
+                // if (#if)
+                TestBase.Diagnostic(ErrorCode.ERR_ExpressionExpected, "").WithLocation(1, 9),
+                // (1,9): error CS1002: ; expected
+                // if (#if)
+                TestBase.Diagnostic(ErrorCode.ERR_SemicolonExpected, "").WithLocation(1, 9),
+                // (1,9): error CS1027: #endif directive expected
```

Comment:
> // (1,9): error CS1027: #endif directive expected [](http://example.com/codeflow?start=15&length=50)

This error feels unexpected. There are similar unexpected  changes to other tests in commit 8.

Reply:
This is the same set of diagnostics reported today.

Before commit 8, we missed some diagnostics because we parsed the directive as BadDirective. In commit 8, we parse it again as normal `#if` directive, just put it into skipped tokens afterwards, but the diagnostics are the same.

Reply:
> In commit 8, we parse it again as normal `#if `directive, just put it into skipped tokens afterwards

Calling the parse helper is an implementation detail. The tree reflects how we parsed the directive and we don't have the directive in the tree. Effectively, that means that the tokens were simply skipped (the mechanism used to skip them is not important), not treated as a directive. Therefore, it is unexpected and undesired (in my opinion) to produce any diagnostics that might imply otherwise.

> but the diagnostics are the same.

We change parsing result, diagnostics should change accordingly, I think.